### PR TITLE
migrate EIP144

### PR DESCRIPTION
**Pull request by lcmmhcc**
- Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/okex/okexchain/blob/develop/CONTRIBUTING.md#pr-targeting))

- [ ] Linked to github-issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Wrote tests
- [ ] Updated relevant documentation (`docs/`)
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Reviewed `Files changed` in the github PR explorer

______

For Admin Use:
- Added appropriate labels to PR (ex. wip, ready-for-review, docs)
- Reviewers Assigned
- Squashed all commits, uses message "Merge PR #XYZ: [title]" ([coding standards](https://github.com/okex/coding/blob/master/README.md#merging-a-pr))

### CODE REVIEW

File: app/rpc/namespaces/eth/api.go [hunk 1]
```diff
@@ -843,12 +853,14 @@ func (api *PublicEthereumAPI) Call(args rpctypes.CallArgs, blockNrOrHash rpctype
...
-	api.addCallCache(key, data.Ret)
+	if overrides == nil {
+		api.addCallCache(key, data.Ret)
+	}
 	return data.Ret, nil
 }
 
 // MultiCall performs multiple raw contract call.
-func (api *PublicEthereumAPI) MultiCall(args []rpctypes.CallArgs, blockNr rpctypes.BlockNumber, _ *map[common.Address]rpctypes.Account) ([]hexutil.Bytes, error) {
+func (api *PublicEthereumAPI) MultiCall(args []rpctypes.CallArgs, blockNr rpctypes.BlockNumber, overrides []*evmtypes.StateOverrides) ([]hexutil.Bytes, error) {
```
Comment (giskook) [inline app/rpc/namespaces/eth/api.go]: maybe use "_" more suitable.

File: app/rpc/namespaces/eth/api.go [hunk 2]
```diff
@@ -946,11 +967,18 @@ func (api *PublicEthereumAPI) doCall(
 	if err != nil {
 		return nil, err
 	}
-
+	simulateQueryData := sdk.QuerySimulateData{
```
Comment (giskook) [inline app/rpc/namespaces/eth/api.go]: how about rename QuerySimulateData to TxStateAccessor or some else? QuerySimulateData seems too general. we can refer this doc: https://geth.ethereum.org/docs/rpc/ns-eth

File: libs/cosmos-sdk/types/context.go [hunk 3]
```diff
@@ -131,6 +133,10 @@ func (c *Context) GetToAccountCacheGas() Gas {
 	return c.accountCache.ToAccGotGas
 }
 
+func (c Context) OverrideBytes() []byte {
```
Comment (giskook) [inline libs/cosmos-sdk/types/context.go]: can we use pointer type as receiver?

File: x/evm/types/api.go [hunk 4]
```diff
@@ -66,3 +70,67 @@ func FormatLogs(logs []vm.StructLog) []*StructLogRes {
...
+		// Override account(contract) code.
+		if account.Code != nil {
+			state.SetCode(addr, *account.Code)
+		}
+		// Override account balance.
+		if account.Balance != nil {
+			state.SetBalance(addr, (*big.Int)(*account.Balance))
+		}
+		if account.State != nil && account.StateDiff != nil {
+			return fmt.Errorf("account %s has both 'state' and 'stateDiff'", addr.Hex())
```
Comment (giskook) [inline x/evm/types/api.go]: if we return errors from here, do the above setted matters ?

Comment (lcmmhcc) [reply x/evm/types/api.go]: no, actually,  theres  a Check() function called before  the Apply function
it'll be better to remove these lines in Apply function, its unnecessary.

File: app/rpc/namespaces/eth/api.go (hunk 2 above)
Comment (lcmmhcc) [reply app/rpc/namespaces/eth/api.go]: this QuerySimulateData struct is used to pass the params from any caller to query the ABCI path "app/simulate". not only for this case.  Maybe it's better to change to SimulateData{}

**Review by giskook — DISMISSED** [review]

**Review by ItsFunny — DISMISSED** [review]

File: libs/cosmos-sdk/baseapp/abci.go [hunk 5]
```diff
@@ -391,7 +392,15 @@ func handleQueryApp(app *BaseApp, path []string, req abci.RequestQuery) abci.Res
 	if len(path) >= 2 {
 		switch path[1] {
 		case "simulate":
-			txBytes := req.Data
+			queryBytes := req.Data
+			var queryData types.SimulateData
+			var txBytes []byte
+			if err := json.Unmarshal(queryBytes, &queryData); err != nil {
```
Comment (ilovers) [inline libs/cosmos-sdk/baseapp/abci.go:434]: Use a new case here? Such as simulateWithState, and only decode json in the new case for better performance.

**Review by ilovers — DISMISSED** [review]

**Review by KamiD — APPROVED** [review]

**Pull request merged**